### PR TITLE
Services 665

### PR DESCRIPTION
**Pull request by sqreek**
This PR intoduce change PUT into PATCH.
because PATCH need to handle modification data (insert, delete, update) of resource, this introduce PreferenceOperation. @Wikia/services-team what do you think about it.

This change follow up changes in wikicitiesDao, that expose modify data that upsert and delete data in one transaction.

Work to do:
- [x] refactor old tests
- [X] write new tests
- [ ] check swagger for changes
- [x] pay technical debt in touched files

### CODE REVIEW

File: service/user-preference/src/test/java/com/wikia/userpreference/helpers/PreferenceTestHelper.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+import com.wikia.userpreference.dao.WikiCitiesDAO;
+import com.wikia.userpreference.domain.Preference;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class PreferenceTestHelper {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/helpers/PreferenceTestHelper.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.io.UnsupportedEncodingException, java.sql.SQLException](https://github.com/Wikia/pandora/blob/6e6271c1107e83bdf70d3fec1bca2197cb873ce3/service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java#L34) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.io.UnsupportedEncodingException, java.sql.SQLException](https://github.com/Wikia/pandora/blob/6e6271c1107e83bdf70d3fec1bca2197cb873ce3/service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java#L51) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)

**Comment by sqreek** [comment]
ping @Wikia/services-team 
related to https://github.com/Wikia/pandora/pull/411

Reopened PR, after missclick and merge previous one

File: service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java [hunk 2]
```diff
@@ -103,7 +103,7 @@ public void getUserPreferencesForValidUserWithoutAuthenticateHeader()
 
   @Test
   public void getUserPreferencesForInvalidUserWithAuthenticateHeader()
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java]: MAJOR Refactor this method to throw at most one checked exception instead of: java.sql.SQLException, java.io.IOException [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 3]
```diff
@@ -1,61 +1,61 @@
...
+        new MockWikiCitiesDataProvider(createMockedUserPropertiesResult(recordList));
+    MockConnection connection = new MockConnection(provider);
+    JooqConfigs database = Mockito.mock(JooqConfigs.class);
+    Mockito.when(database.dslForReads(WikiCitiesDAO.DB_NAME)).thenReturn(DSL.using(connection));
+
+    return database;
+  }
+
+  @SuppressWarnings("unchecked")
+  public static Result<Record> createMockedUserPropertiesResult(ArrayList<Record> recordsList) {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MAJOR The type of the "recordsList" object should be an interface such as "List" rather than the implementation "ArrayList". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1319)

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 4]
```diff
@@ -1,61 +1,61 @@
...
+import org.jooq.Result;
+import org.jooq.SQLDialect;
+import org.jooq.impl.DSL;
+import org.jooq.tools.jdbc.MockConnection;
+import org.jooq.tools.jdbc.MockDataProvider;
+import org.mockito.Mockito;
+
+import java.util.ArrayList;
+
+public class JooqMockHelper {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 5]
```diff
@@ -1,61 +1,61 @@
...
+
+  public static JooqConfigs createWikiCitiesDataSourceUsingWikiCitiesDataProvider(
+      ArrayList<Record> recordList) throws java.sql.SQLException {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MAJOR The type of the "recordList" object should be an interface such as "List" rather than the implementation "ArrayList". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1319)

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 6]
```diff
@@ -1,61 +1,61 @@
...
+
+  @SuppressWarnings("unchecked")
+  public static Result<Record> createMockedUserPropertiesResult(ArrayList<Record> recordsList) {
+    DSLContext create = DSL.using(SQLDialect.MYSQL);
+    Field propertyField = DSL.fieldByName("up_property");
+    Field valueField = DSL.fieldByName("up_value");
+    Field userIdField = DSL.fieldByName("up_user");
+    Result<Record> result = create.newResult(propertyField, valueField, userIdField);
+    recordsList.stream().forEach(
+        (record) -> {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MINOR Remove the parentheses around the "record" parameter [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1611)
MAJOR Remove useless curly braces around statement [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1602)

File: service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java [hunk 7]
```diff
@@ -17,10 +19,14 @@
 import com.wordnik.swagger.annotations.ApiResponses;
 import com.wordnik.swagger.annotations.Authorization;
 import io.dropwizard.auth.Auth;
+import io.dropwizard.jersey.PATCH;
 
 import java.sql.SQLException;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java]: MINOR Remove this unused import 'java.util.Optional'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+  public static final String UTF_8 = "UTF-8";
+  public static final String PROPERTY = "property";
+  public static final String PROPERTY_1 = "property1";
+  public static final String VALUE = "value";
+  public static final String VALUE_1 = "value1";
+
+
+  @Test
+  public void testGetAllUserPreferences() throws UnsupportedEncodingException, SQLException {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java]: MAJOR Refactor this method to throw at most one checked exception instead of: java.io.UnsupportedEncodingException, java.sql.SQLException [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)

File: service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    WikiCitiesDAO wikiCitiesDAO = new WikiCitiesDAO(database, new Whitelist());
+
+    Result<Record> result = wikiCitiesDAO.getWhiteListedUserPreferences(VALID_USER_ID);
+    assertTrue(result.isNotEmpty());
+    assertEquals(1, result.size());
+    Mockito.verify(database).dslForReads(WikiCitiesDAO.DB_NAME);
+  }
+
+  @Test
+  public void testResultToMap() throws UnsupportedEncodingException, SQLException {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOTest.java]: MAJOR Refactor this method to throw at most one checked exception instead of: java.io.UnsupportedEncodingException, java.sql.SQLException [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 16 issues:
* MAJOR 14 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.sql.SQLException, java.net.URISyntaxException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L80) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.sql.SQLException, java.io.IOException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java#L65) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.sql.SQLException, java.io.IOException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java#L146) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.sql.SQLException, java.io.IOException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java#L135) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.sql.SQLException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java#L159) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)
* MAJOR [Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.sql.SQLException](https://github.com/Wikia/pandora/blob/e25caae4528f71ff0c8460a3cfa60b4407e4a905/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceTest.java#L118) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1160)

**Comment by nmonterroso** [comment]
I don't think we can just change the current `put` operation to `patch` since Helios depends on the current implementation of `put`. If we make this change and push it out, Helios would not be able to set preferences. 

Also, `WikiCities` and `WikiCitiesDAO` are deprecated - we are only keeping them while we transition Helios to use resources that use the `StoredPreferences` interface, which has a subtask in this sprint [here](https://wikia-inc.atlassian.net/browse/SERVICES-741).

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 10]
```diff
@@ -1,61 +1,62 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class JooqMockHelper {
+  private JooqMockHelper() {
+
+  }
+
+  public static JooqConfigs createWikiCitiesDataSourceUsingWikiCitiesDataProvider(
+      ArrayList<Record> recordList) throws java.sql.SQLException {
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MAJOR The type of the "recordList" object should be an interface such as "List" rather than the implementation "ArrayList". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1319)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java [hunk 11]
```diff
@@ -1,61 +1,62 @@
...
+import org.jooq.Field;
+import org.jooq.Record;
+import org.jooq.Result;
+import org.jooq.SQLDialect;
+import org.jooq.impl.DSL;
+import org.jooq.tools.jdbc.MockConnection;
+import org.jooq.tools.jdbc.MockDataProvider;
+import org.mockito.Mockito;
+
+import java.util.ArrayList;
```
Comment (WikiaTech) [inline service/user-preference/src/test/java/com/wikia/userpreference/util/JooqMockHelper.java]: MINOR Remove this unused import 'java.util.ArrayList'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by sqreek** [comment]
@nmonterroso, but you already changed the [logic of PUT]
(https://github.com/Wikia/pandora/blob/SERVICES-665/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L107)

a) helios do this on user creation, so user don't had any preference to override but 
b) it still need to be changed in helios.

**Comment by nmonterroso** [comment]
@sqreek yes we [changed](https://github.com/Wikia/pandora/blob/master/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L102) it but we [kept the original](https://github.com/Wikia/pandora/blob/master/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L133) so for now the preference resource accepts both `PUT /{userId} Accept: application/v2+json` for the new-style and `PUT /{userId}` to support what Helios already does. We did this so we didn't introduce a breaking change.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "No user preference found" 3 times.](https://github.com/Wikia/pandora/blob/b395062610dfaf5d967dcdce36878f7bebf64fed/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L123) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1192)

**Comment by sqreek** [comment]
ok, So I revert old setUserPreferencesLegacy as PUT, and leave PATCH there.
Is it ok?

**Comment by nmonterroso** [comment]
Yah, that sounds good :)

Though the new `PATCH` would have to accept `GlobalPreference` and `LocalPreference` objects, not the soon-to-be-deprecated `Preference`

**Comment by sqreek** [comment]
Ok, so if there is no blocker for now (except accept for GlobalPreference and LocalPreference), I want to merge this partial of work to not blocking other. @drsnyder for example.
ping @nmonterroso .
I will end this tommorow with add Local and Global Preference for PATCH

**Comment by drsnyder** [comment]
@sqreek the work associated with this ticket is not in the current sprint.  I think we should defer this until we need the `PATCH` support.

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 12]
```diff
@@ -166,4 +210,35 @@ public boolean setUserPreference(Integer userId, String prefKey, String newValue
...
+          if (preferenceWhiteList.allows(preference.getName())) {
+            whiteListedPreferences.add(preference);
+          } else {
+            invalidPreferenceNames.add(preference.getName());
+          }
+        }
+      }
+    }
+
+    public boolean containsInvalidPreferenceNames() {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: containsInvalidPreferenceNames  -> hasInvalidPreferenceNames?
contains is more like lookup thing than test if not empty.

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 13]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
-                  .build()
-          ),
-          "Non-whitelisted preferences were attempted to be set");
+    if (filteredPreferences.containsInvalidPreferenceNames()) {
+      logInvalidPreferenceNames(filteredPreferences.invalidPreferenceNames);
     }
 
-    if (preferencesToUpdate.isEmpty() && !invalidPreferenceNames.isEmpty()) {
+    if (filteredPreferences.whiteListedPreferences.isEmpty() &&
+        !filteredPreferences.invalidPreferenceNames.isEmpty()) {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: You should use `filteredPreferences.containsInvalidPreferencesNames()`. I think you can just:
```java
else if (filteredPreferences.isEmpty()) {
}
```

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 14]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
+            userId,
+            preference.getName().getBytes(),
+            preference.getValue().getBytes()));
+  }
+
+  private InsertValuesStep3<UserPropertiesRecord, Integer, byte[], byte[]> createInsertUserPropertiesStatement() {
+    return createInsertUserPropertiesStatement(database.dslForWrites(DB_NAME));
+  }
+
+  private InsertValuesStep3<UserPropertiesRecord, Integer, byte[], byte[]> createInsertUserPropertiesStatement(DSLContext context) {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: Why this method exists? This could be moved to the one above and would works just fine?

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 15]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
         .onDuplicateKeyUpdate()
-        .set(
-            field(USER_PROPERTIES.UP_VALUE.getName()),
+        .set(field(USER_PROPERTIES.UP_VALUE.getName()),
             (Object) field("VALUES(" + USER_PROPERTIES.UP_VALUE.getName() + ")"));
+    return statement;
+  }
 
-    int i = statement.execute();
+  private void logInvalidPreferenceNames(List<String> invalidPreferenceNames) {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: I wonder if that shouldn't be a part of `FilteredPreferences` class, it seems it is always invoked in that context.

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 16]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
+
+    logInvalidPreferenceNames(filteredPreferencesToUpsert.invalidPreferenceNames);
+    logInvalidPreferenceNames(filteredPreferencesToDelete.invalidPreferenceNames);
+
+    final Integer[] modifiedElements = new Integer[1];
+    database.dslForWrites(DB_NAME).transaction(configuration -> {
+      InsertValuesStep3<UserPropertiesRecord, Integer, byte[], byte[]> upsertStatement =
+          createInsertUserPropertiesStatement(DSL.using(configuration));
+      addPreferencesIntoStatement(userId,
+          filteredPreferencesToUpsert.whiteListedPreferences,
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: Should we check if there are any preferences to save?

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 17]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
+
+  private void addPreferencesIntoStatement(Integer userId, List<Preference> preferences, InsertValuesStep3<UserPropertiesRecord, Integer, byte[], byte[]> statement) {
+    preferences.forEach(preference ->
+        statement.values(
+            userId,
+            preference.getName().getBytes(),
+            preference.getValue().getBytes()));
+  }
+
+  private InsertValuesStep3<UserPropertiesRecord, Integer, byte[], byte[]> createInsertUserPropertiesStatement() {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: Wy bother creating this proxy?

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 18]
```diff
@@ -80,61 +79,106 @@ public static String byteArrayToString(byte[] array) {
...
+          createInsertUserPropertiesStatement(DSL.using(configuration));
+      addPreferencesIntoStatement(userId,
+          filteredPreferencesToUpsert.whiteListedPreferences,
+          upsertStatement);
+      int upsertElements = upsertStatement.execute();
+      Delete deleteStatement = DSL.using(configuration)
+          .delete(USER_PROPERTIES)
+          .where(USER_PROPERTIES.UP_USER.equal(userId))
+          .and(USER_PROPERTIES.UP_PROPERTY.in(
+              filteredPreferencesToDelete.whiteListedPreferences
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: again we should check for empty lists

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 19]
```diff
@@ -166,4 +210,35 @@ public boolean setUserPreference(Integer userId, String prefKey, String newValue
...
+
+    public FilteredPreferences(List<Preference> preferences) {
+      invalidPreferenceNames = new ArrayList<>();
+      if (preferences == null) {
+        whiteListedPreferences = new ArrayList<>();
+        return;
+      }
+
+      if (preferenceWhiteList.isEmpty()) {
+        whiteListedPreferences = preferences.stream().collect(Collectors.toList());
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: copy/clone? Why using streams here?

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 20]
```diff
@@ -166,4 +210,35 @@ public boolean setUserPreference(Integer userId, String prefKey, String newValue
 
     return statement.fetch();
   }
+
+  private class FilteredPreferences {
+    private List<Preference> whiteListedPreferences;
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: Since it is a list how we manage duplicates?

File: service/user-preference/src/main/java/com/wikia/userpreference/domain/UserPreferences.java [hunk 21]
```diff
@@ -66,4 +66,20 @@ public boolean hasLocalPreferences() {
...
+            .filter(p -> p.getName().equals(preference))
+            .count() == 1;
+  }
+
+  public boolean containsLocalPreference(String preference, int wikiId) {
+    return hasLocalPreferences() &&
+        getLocalPreferences()
+            .stream()
+            .filter(p -> p.getName().equals(preference) && p.getWikiId() == wikiId)
+            .count() == 1;
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/domain/UserPreferences.java]: Here you assume that local preference list has unique preferences.

File: service/user-preference/src/main/java/com/wikia/userpreference/domain/PreferencePathOperation.java [hunk 22]
```diff
@@ -0,0 +1,11 @@
+package com.wikia.userpreference.domain;
+
+public enum PreferencePathOperation {
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/domain/PreferencePathOperation.java]: `PreferencePathOperation` -> `PreferencePatchOperation`?

File: service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java [hunk 23]
```diff
@@ -150,9 +155,57 @@ public Response setUserPreferencesLegacy(
...
+
+
+    boolean modifyUserPreferenceResult =
+        wikiCities.modifyUserPreferences(
+            userId,
+            preferenceOperationMap.get(PreferencePathOperation.UPSERT),
+            preferenceOperationMap.get(PreferencePathOperation.DELETE));
+
+    if (modifyUserPreferenceResult) {
+      return StandardResponses.Successful("Successful modify user preferences");
```
Comment (harnash) [inline service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java]: `"Successful modify user preferences"` -> `"User preferences successfully modified"`

File: service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOIntegrationTest.java [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+    Optional<List<Preference>> preferences = wikiCitiesDAO.getUserPreferences(VALID_USER_ID);
+
+    assertTrue(preferences.isPresent());
+    for (String propertyName : literals) {
+      assertTrue(
+          preferences
+              .get()
+              .stream()
+              .filter(p -> p.getName().equals(propertyName))
+              .count() > 0);
```
Comment (harnash) [inline service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOIntegrationTest.java]: We are assuming that user has valid properties. Shouldn't we just fetch the properties without whitelisting and then compare the results?

File: service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOIntegrationTest.java [hunk 25]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    for (String regex : regexes) {
+      Pattern p = Pattern.compile(regex);
+      assertTrue(
+          "no matches found for " + regex,
+          preferences.get().stream()
+              .filter(pref -> p.matcher(pref.getName()).matches())
+              .count() > 0);
+    }
```
Comment (harnash) [inline service/user-preference/src/test/java/com/wikia/userpreference/dao/WikiCitiesDAOIntegrationTest.java]: Here we check that whitelisted properties exists (they might not be set at all on the given user. We do not check if function returned properties that are not whitelisted.

**Comment by sqreek** [comment]
https://github.com/Wikia/pandora/pull/460

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "No user preference found" 3 times.](https://github.com/Wikia/pandora/blob/b395062610dfaf5d967dcdce36878f7bebf64fed/service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceResource.java#L123) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1192)
